### PR TITLE
Added Upcoming workshop API code for SOMRC Website

### PR DESCRIPTION
**Pull request by orksomrc**

### CODE REVIEW

File: backend/api-processor/workshop_processor.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#!/bin/python
```
Comment (danfunk) [inline backend/api-processor/workshop_processor.py]: You should remove this file, we won't be using it, so should not be part of the pull request.

File: backend/.env [hunk 2]
```diff
@@ -1,2 +1,3 @@
 source activate edp
```
Comment (danfunk) [inline backend/.env]: there are no actual changes in this file, so this should not be a part of the pull request.

File: backend/ed_platform/discourse.py [hunk 3]
```diff
@@ -16,8 +16,9 @@ def __init__(self, app):
         self.key = app.config['DISCOURSE']['key']
```
Comment (danfunk) [inline backend/ed_platform/discourse.py]: This is a breaking change.  I know we had to do this to get it working for you locally, but this should not be a part of the pull request.

File: backend/ed_platform/views.py [hunk 4]
```diff
@@ -256,6 +260,49 @@ def set_track_image(id):
     db.session.commit()
     return get_track_image(id)
 
+# Latest Workshops
```
Comment (danfunk) [inline backend/ed_platform/views.py]: You should remove these comments.

File: backend/ed_platform/views.py [hunk 5]
```diff
@@ -7,6 +7,10 @@
 from ed_platform import app, db, models, sso, RestException, elastic_index, profile_photos, discourse, notify
 from flask_httpauth import HTTPTokenAuth
 
+import requests, json, yaml
```
Comment (danfunk) [inline backend/ed_platform/views.py]: Most of these imports do not seem necessary.

File: backend/ed_platform/views.py [hunk 6]
```diff
@@ -7,6 +7,10 @@
...
+from dateutil import parser
+from operator import itemgetter
```
Comment (danfunk) [inline backend/ed_platform/views.py]: I don't think this is used.

File: backend/ed_platform/views.py [hunk 7]
```diff
@@ -268,6 +315,40 @@ def get_workshops():
                          models.Workshop.query.all()))
     return jsonify({"workshops": workshops})
 
+@app.route('/api/upcoming-workshops')
+def get_upcoming_workshops():
+    trimmed_workshop_list = []
+    date_today = datetime.datetime.now()
+    date_today - date_today.replace(tzinfo=None)
```
Comment (danfunk) [inline backend/ed_platform/views.py]: Does this line do anything? That's a "-" sign, not an equals sign.

File: backend/ed_platform/views.py [hunk 8]
```diff
@@ -268,6 +315,40 @@ def get_workshops():
...
+    workshop_data = list(map(lambda t: workshop_schema.dump(t).data,
+                         models.Workshop.query.all()))
+    for workshop in workshop_data['workshops']:
+        if (len(workshop['sessions']) > 0):
+            id = workshop['id']
+            title_str = workshop['title']
+            display_name_str = workshop['instructor']['display_name']
+            session_date = parser.parse(workshop['sessions'][0]['date_time'])
+            session_date = session_date.replace(tzinfo=None)
+            workshop_dict = {
```
Comment (danfunk) [inline backend/ed_platform/views.py]: We are using Marshmallow Schema files to generate Json.  You don't need to create these dictionaries manually.  If we consistently use the Schema files then we produce consistent Json structures for the front end.  I would create a list of Workshop models that meet the requirements, then covert that list of workshops to json with a line like:

models.WorkshopAPISchema().jsonify(**MyWorkshopList**,many=True)

File: backend/migrations/versions/df361a922e41_.py [hunk 9]
```diff
@@ -28,6 +28,10 @@ def upgrade():
     # ### commands auto generated by Alembic - please adjust! ###
     op.add_column('participant', sa.Column('new_account', sa.Boolean(), nullable=True))
     op.add_column('participant', sa.Column('use_gravatar', sa.Boolean(), nullable=True))
+    # ### end Alembic commands ###
```
Comment (danfunk) [inline backend/migrations/versions/df361a922e41_.py]: We removed this at some point because it was breaking new migrations.  Not sure why it is showing back up in this pull request, but it shouldn't be here.

File: backend/tests.py [hunk 10]
```diff
@@ -8,7 +8,7 @@
 import requests
 from io import StringIO, BytesIO
 from flask import json, request
-from flask_mail import Mail
+# from flask_mail import Mail
```
Comment (danfunk) [inline backend/tests.py]: Why is this commented out?

File: backend/tests.py [hunk 11]
```diff
@@ -236,6 +236,15 @@ def get_workshop(self, id):
                           content_type="application/json")
         return json.loads(rv.get_data(as_text=True))
 
+    def get_latest_workshops(self):
```
Comment (danfunk) [inline backend/tests.py]: This is a good sanity check test. Consider adding some additional tests that match the requirements.  Should we only ever get six recommended workshops back?  Should those workshops be the latest workshops available? etc...

File: frontend/src/app/workshop-details/workshop-details.component.html [hunk 12]
```diff
@@ -103,7 +103,7 @@ <h2>Sessions</h2>
 
     <div class="discourse" *ngIf="workshop.discourse_enabled && isDiscourseLoaded">
       <h2>Posts in Discussion</h2>
-      <p>You can participate in this discussion on <a target="discourse"
```
Comment (danfunk) [inline frontend/src/app/workshop-details/workshop-details.component.html]: Good catch!  Thank you.

**Comment by orksomrc** [comment]
Test email body

—
Omar R. Khurshid
Cloud DevOps Engineer
UVa School of Medicine Research Computing
Need to set up HIPAA-compliant cloud infrastructure? We can help:
https://somrc.virginia.edu<https://somrc.virginia.edu/>
[id:image001.png@01D39105.CBB1AF80]
